### ISSUE TITLE
`@nograd` pullbacks cannot be overloaded with `@adjoint` (`Context` vs. `AContext`)

### ISSUE THREAD

**Issue by laggvar** [FluxML/Zygote.jl#1144]
I have a custom type for which a corresponding method of `Base.trunc` makes sense to differentiate. But https://github.com/FluxML/Zygote.jl/blob/169563f65f0d9f9221ad528780aab48d157a05fc/src/lib/number.jl#L2 means `@adjoint` does not work, because

```
_pullback(::Context, typeof(trunc), args::Any...) #Created by @nograd
_pullback(::AContext, typeof(trunc), t::MyType, args::Any...) # Created by @adjoint
```
are ambiguous. While there are many ways in which this could be solved, the fundamental problem seems to be `@nograd` creating methods for the concrete type `Context`. 

I cannot find why some methods do not dispatch on the abstract type (and it does not seem very important in a previous issue https://github.com/FluxML/Zygote.jl/pull/844), but perhaps there is a good reason here?

**Comment by willtebbutt** [comment]
We've moved this forwards a bit in ChainRules -- see the `@non_differentiable` macro. Probably the right solution is to remove these `@nograd`-ed things from Zygote, and `@non_differentiable` them more selectively in ChainRules.

**Comment by laggvar** [comment]
Thank you!

Gave it a shot. Overloads seem to behave as expected if defined this way 👍 
```
julia> gradient(x -> myfunc(x, 1)[1], MyType())
# Angry error
###########
julia> ChainRulesCore.@non_differentiable myfunc(args...)
julia> gradient(x -> myfunc(x, 1)[1], MyType())
(nothing, )
###########
julia> ChainRulesCore.@non_differentiable myfunc(args...)
julia> ChainRulesCore.rrule(typeof(myfunc), t::MyType, args...) = <code>
julia> gradient(x -> myfunc(x, 1)[1], MyType())
<correct output> 
```
So, yes, this looks like the good solution. 

Have I understood correctly that ChainRules will only be used if Zygote has no `_pullback` method for the generic fallback in interface2.jl? https://github.com/FluxML/Zygote.jl/blob/169563f65f0d9f9221ad528780aab48d157a05fc/src/compiler/interface2.jl#L9
I.e. things defined in Zygote itself will always shadow ChainRules?

ChainRules seem to not implement anything for `div` and `trunc`, but `floor`, `ceil` and `round` are unfortunately already  implemented to return `zero(x)` instead of `nothing`
https://github.com/JuliaDiff/ChainRules.jl/blob/68532f25f85533202dc616f09d7547cc3d53a1ef/src/rulesets/Base/base.jl#L178

**Comment by mcabbott** [comment]
Yes to Zygote taking preference. But any rule here can be removed here once it's moved to CR.

> return zero(x) instead of nothing

Not sure why those are like that. The logic of `zero(x)` not having gradient `nothing` is that a common use of `zero` is to make functions like `f(x) = x<3 ? zero(x) : x^2` type-stable, and this makes the gradient also stable. But it's hard to see `trunc` being used that way. Maybe it was discussed on making that rule, or maybe not.

**Comment by DhairyaLGandhi** [comment]
It should still be possible to achieve this via overloading `_pullback`, so there's no fundamental limitation in this as such.

**Comment by ToucheSir** [comment]
Like `@non_differentiable`, `@scalar_rule` generates rule overloads behind the scenes. You can look at the output of `@macroexpand @scalar_rule trunc(x) zero(x)` for example. Because of this, I believe the rule for the custom type shown above should also override the one from `@scalar_rule`.

**Comment by CarloLucibello** [comment]
`@nograd` is deprecated and `trunc`/`floor`/etc. are now handled via ChainRules `@non_differentiable`, so the old `@nograd`-vs-`@adjoint` Context conflict is gone; a custom `rrule` works without ambiguity. Closing as fixed.

_(Triaged against Zygote v0.7.10 on Julia 1.12.6.)_
